rfd/0033-desktop-access.md

```diff
@@ -0,0 +1,176 @@
+record the session.
+
+In Desktop Access, the client is a web application, bundled as part of the
+Teleport Web UI (similar to the web SSH client). The backend service (e.g.
+`windows_desktop_service`) is OS-specific and can represent multiple desktop
+hosts.
+
+```
++--------+
+| web UI |
++--------+
+    ^
+    | desktop protocol over websocket
+    v
++--------+
+| proxy  |
++--------+
+    ^
+    | desktop protocol over mTLS
+    v
++---------+
+| backend |
+| service |<----------------+
++---------+                 |
+   ^                        |
+   | OS protocol (e.g. RDP) |
+   v                        v
++---------+              +---------+
+| desktop |              | desktop |
+| host 1  |              | host 2  |
++---------+              +---------+
+```
+
+From web UI to the backend service, the underlying wire protocol is a custom
+protocol, which we'll call "desktop protocol". Between the backend service and
+the actual desktop hosts, the wire protocol is a standard protocol for the
+target OS (like RDP or X11), which we'll call "OS protocol".
+
+### Protocol
+
+The wire protocol used between the Teleport UI and backend service is a custom
+binary protocol. It is described in more detail in RFD 37.
+
+We are not using any standard remote desktop protocol, like
+RDP/X11/VNC/Guacamole for two reasons:
+- interoperability with 3rd party software is not a requirement
+- avoid legacy baggage and complexity of standard protocols, focus on
+  simplicity and performance
+
+We are also not using generic encoding protocols, like gRPC, to keep
+client-side parsing simple and reduce the JS bloat.
+
+### Client
+
+The client is a web app built into the Teleport Web UI. It's a JavaScript/TypeScript application that:
+- picks one of the allowed OS usernames for the logged-in Teleport user based
+  on their roles
+- creates a websocket connection
+- handles the incoming desktop protocol messages and renders them onto an HTML
+  canvas
+- captures user input, encodes it into desktop protocol messages and sends them
+  over the websocket
+
+TODO: wireframes?
+
+### Configuration
+
+There are no additional configuration options for Desktop Access in the web UI
+or the proxy. The Desktop Access client is part of the web UI and uses the same
+web port on the proxy for connecting a websocket.
+
+All relevant configuration to enable the feature will live in OS-specific
+backend services, described in their own RFDs (like RFD 34 for Windows).
+
+### Authorization
+
+Role definitions for RBAC have a few new fields for Desktop Access:
+- `${OS}_desktop_logins` - list of desktop login names allowed/denied for
+  desktop hosts with a given OS
+- `${OS}_desktop_labels` - list of labels to match the target desktop hosts
+  against
+- `desktop_clipboard` - option for allowing copy/paste to/from the desktop
+  host; default is `true`
+- `record_desktop_session` - option for enabling/disabling recording of
```

Comment:
Would there be any other way to enable/disable session recording outside of the role configuration? Will this model allow us to support configurations such as "always record sessions in the prod environment, but do not record in the dev environment?"

Reply:
That's a good question.
You can achieve the example configuration you gave by creating prod- and dev-specific roles, which seems like a good practice anyway.

We could add some additional recording configuration at the `windows_desktop_service` level, but it would make the rules of when recording happens complex. Configuring teleport is already challenging for most users.

I'd like to stick with our current RBAC-centric model unless there's a high demand for something more flexible